lld/MachO/MapFile.cpp

```diff
@@ -217,15 +203,42 @@ void macho::writeMapFile() {
                    seg->name.str().c_str(), osec->name.str().c_str());
     }
 
-  // Shared function to print an array of symbols.
-  auto printIsecArrSyms = [&](const std::vector<ConcatInputSection *> &arr) {
-    for (const ConcatInputSection *isec : arr) {
+  // Shared function to print one or two arrays of ConcatInputSection in
+  // ascending outSecOff order. The second array is optional; if provided, we
+  // interleave the printing in sorted order without allocating a merged temp
+  // array.
+  auto printIsecArrSyms = [&](const std::vector<ConcatInputSection *> &arr1,
+                              const std::vector<ConcatInputSection *> *arr2 =
+                                  nullptr) {
+    // Helper lambda that prints all symbols from one ConcatInputSection.
+    auto printOne = [&](const ConcatInputSection *isec) {
       for (Defined *sym : isec->symbols) {
-        if (!(isPrivateLabel(sym->getName()) && getSymSizeForMap(sym) == 0))
+        if (!(isPrivateLabel(sym->getName()) && getSymSizeForMap(sym) == 0)) {
           os << format("0x%08llX\t0x%08llX\t[%3u] %s\n", sym->getVA(),
                        getSymSizeForMap(sym),
-                       readerToFileOrdinal[sym->getFile()],
+                       readerToFileOrdinal.lookup(sym->getFile()),
                        sym->getName().str().data());
+        }
+      }
+    };
+
+    // If there is only one array, print all symbols from it and return.
+    // This simplifies the logic for the merge case below.
+    if (!arr2) {
+      for (const ConcatInputSection *isec : arr1)
+        printOne(isec);
+      return;
+    }
+
+    size_t i = 0, j = 0;
+    size_t size1 = arr1.size();
+    size_t size2 = arr2->size();
+    while (i < size1 || j < size2) {
+      if (i < size1 &&
+          (j >= size2 || arr1[i]->outSecOff <= (*arr2)[j]->outSecOff)) {
```

Comment:
I agree with the general logic, but `(*arr2)[j]` is not so please to look at, why not `ArrayRef`? IIRC you can have `ArrayRef` pointing to `nullptr`s and it would be prettier.